CodeGeneration/Sources/SyntaxSupport/CommonNodes.swift

```diff
@@ -180,13 +180,98 @@ public let COMMON_NODES: [Node] = [
     parserFunction: "parseDeclaration"
   ),
 
+  Node(
+    kind: .declGroupHeader,
```

Comment:
None of the other base nodes (`DeclSyntax`, `ExprSyntax` etc.) have any children and I think we will get crashes if any of the nodes whose base is `.declGroupHeader` are layout-incompatible with `declGroupHeader`. `DeclGroupHeaderSyntax` will assume that the first child in the node is `attributes` but that might not be the case for a node inheriting from `.declGroupHeader`. 

That said, I would make the base of all header nodes `.syntax` and create a trait for the decl group headers instead.

Reply:
I'm a little reluctant to do this because I get a fair bit of mileage out of having `DeclGroupHeaderSyntax` be a concrete type instead of an existential. For instance, both `ABIAttributeSyntax.Provider` and the modified version of `HasTrailingMemberBlock` use the concrete `DeclGroupHeaderSyntax`; the former could not easily use an existential because there's no way to constrain a child node to have a trait, and the latter because a concrete type is needed to provide a contextual type for the string literal.

(If it helps, the `DeclGroupHeaderSyntax` accessors don't directly access the children—they indirect through the existential so they use the leaf type's layout.)

If you can suggest another design that would still give me a concrete node which could contain any of the headers, I'd be open to that change.

Reply:
Let me think about it some more. It might also become clearer to me once https://github.com/swiftlang/swift-syntax/pull/2888 is merged and this PR is rebased so it doesn’t contain those changes.

Reply:
Hey @ahoppen , we now have the other PR merged. Do you have more specific suggestions in mind?